### PR TITLE
Avoid hard-coding list of Node Conditions

### PR DESCRIPTION
**Pull request by justinsb**
We assume that if a Condition isn't well-known, that it blocks
scheduling of pods, and that the "unhealthy" value is api.ConditionTrue

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36210)

### CODE REVIEW

**Comment by justinsb** [comment]
cc @wlan0

**Comment by dcbw** [comment]
Seems reasonable to me...

**Comment by wlan0** [comment]
+1

**Comment by davidopp** [comment]
LGTM

**Comment by saad-ali** [comment]
Release-czar approved post-code freeze merge--This was LGTMed and in the merge-queue at code freeze time for 1.5. Adding 1.5 milestone to let it gets merged after code freeze.

**Comment by dims** [comment]
@k8s-bot test this

**Comment by dims** [comment]
from https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/36210/pull-kubernetes-unit/4668/build-log.txt
```
W1110 16:15:37.879] # k8s.io/kubernetes/plugin/pkg/scheduler/factory
W1110 16:15:37.879] plugin/pkg/scheduler/factory/factory.go:517: undefined: "k8s.io/kubernetes/pkg/api".NodeInodePressure
I1110 16:16:03.470] process 342 exited with code 2
```

Looks like this PR is not ready. Is this still needed for v1.5 milestone? (if not, can you please switch that?)

**Comment by justinsb** [comment]
Just needed a rebase I think...

@davidopp OK to reapply LGTM?

File: plugin/pkg/scheduler/factory/factory.go [hunk 1]
```diff
@@ -507,15 +507,22 @@ func getNodeConditionPredicate() cache.NodeConditionPredicate {
...
+				if cond.Status != api.ConditionTrue {
+					glog.V(4).Infof("Ignoring node %v with %v condition status %v", node.Name, cond.Type, cond.Status)
+					return false
+				}
+
+			case api.NodeMemoryPressure, api.NodeDiskPressure:
+			// We don't block on "pressure" conditions; these are warnings, not errors!
+
+			default:
+				// We assume everything else is blocking if the condition is True or Unknown
```
Comment (liggitt) [inline plugin/pkg/scheduler/factory/factory.go]: this seems like a bad idea... will all future conditions for all time require a value of False to be schedulable? I'd expect a generic scheduler predicate that lets you specify a condition and required value(s) in order to schedule

**Comment by liggitt** [comment]
This needed API approval before we lock the meaning of Condition values in the API going forward (`the "unhealthy" value is api.ConditionTrue`). I think this needs to be reverted before we ship 1.5 with this semantic. 

If this is in service of custom node conditions blocking scheduling, that could be done with a scheduler predicate that allows requires a specific condition with a specific value (or values). Newly created nodes would be missing the condition entirely (and would not schedule), and an external controller could add the custom condition at its leisure. That would not require an admission plugin (https://github.com/kubernetes/kubernetes/pull/36209) and would not require this change to assume True means unhealthy.

**Comment by dchen1107** [comment]
I commented this pr at: https://github.com/kubernetes/kubernetes/issues/25773#issuecomment-262034716

I think this pr changed the semantics of NodeCondition for today's user and long-term too. We should revert this pr IMHO. cc/ @bgrant0607

**Comment by justinsb** [comment]
It seems that the concern is that if a user is setting a node status condition, that it is no longer ignored?

That feels like us fixing a pre existing bug. Given the status conditions didn't do anything before, do we know of anyone that used them? It is exactly because we want to let these conditions be useful that we need to stop hard coding them 😀

Is a release note appropriate?

**Comment by liggitt** [comment]
> It seems that the concern is that if a user is setting a node status condition, that it is no longer ignored? That feels like us fixing a pre-existing bug.

I didn't think all conditions inherently mapped directly to schedulability. If a user wants to set something custom to affect scheduling, I thought that's what labels/taints were for.

> Given the status conditions didn't do anything before, do we know of anyone that used them?

Custom conditions didn't drive anything in the scheduler, but that doesn't mean we're free to start mapping them to schedulability (and require future conditions to map that way, with fixed values).

> It is exactly because we want to let these conditions be useful that we need to stop hard coding them

This exchanges one form of hard-coding (types+statuses) for another (status without regard for type), and constrains future use of node conditions. To ensure the conditions are present as soon as the Node object exists, it also requires configuring the node to populate custom conditions or configuring the apiserver (via admission options, like https://github.com/kubernetes/kubernetes/pull/36209) to set initial conditions.

If we want custom conditions to affect schedulability, that seems like it needs to be opt-in, and informed by knowledge about the meaning of the presence/absence of the condition, as well as the meaning of the condition's values (e.g. https://github.com/kubernetes/kubernetes/issues/37224). That would also allow adding the conditions asynchronously, removing the need for node/admission modifications.

**Comment by liggitt** [comment]
will wait for @kubernetes/kube-api to weigh in, but since this seems like both a breaking API change and a divergence from the scheduler/conditions/taints approach discussed elsewhere, opened a speculative revert in https://github.com/kubernetes/kubernetes/pull/37234

**Comment by bgrant0607** [comment]
Our plan is to model all forms of unschedulability using taints, but that needs to at least reach beta before we can rely on it.

https://github.com/kubernetes/kubernetes/pull/18263#issuecomment-169220144

We need to find somewhere obvious to document our POR.

**Comment by davidopp** [comment]
I am not as concerned about the requirement that unhealthy is always represented as true, as much as I am about Jordan's good point that this means all unmentioned node conditions actually have to affect schedulability, which does seem like a bad assumption. I agree we should revert this PR, sorry I did not think about these issues when I reviewed it.

We already have an issue open to replace unschedulable with a taint (#29178), and #34825 converts Ready, but we don't have anything explicit about converting the other node conditions. I am not sure whether we will get rid of the conditions, or add taints in addition to the conditions and stop having the conditions affect scheduling -- I'm open to suggestions/opinions though would lean towards keeping the conditions. (Unschedulable is different, it is part of the spec and not a node condition.)

@justinsb was there a specific use case that motivated this PR?

**Comment by bgrant0607** [comment]
@davidopp The conditions are intended to provide additional information to the user, such as transition time, reason, and human-readable message about the cause, so I'd keep them.

**Comment by justinsb** [comment]
The core motivation was this issue: https://github.com/kubernetes/kubernetes/pull/33573

Additional motivation came from the desire to pull stuff out of core (e.g. cloudproviders, as @wlan0 is working on).

I did think that was the design of Node Status Conditions (soon to be moved to taints) - that this would enable us to split out into separate controllers, in multiple repositories.

If I can get #36209 merged I think that solves the issue for me.  I would guess if we made it auto-apply a taint it would solve it for the cloudprovider use case also.

**Comment by liggitt** [comment]
#36209 by itself isn't sufficient, since the scheduler doesn't pay attention to conditions it doesn't know about. You'd need something like #37224 to enable that, and wouldn't actually need the admission plugin in that case

**Comment by justinsb** [comment]
@davidopp  #36209 would be configured when a networking driver was configured that updated the status appropriately (currently just kubenet, but that'll change fast).

If we want to use _new_ Node Status Conditions to separate out logic e.g. cloudproviders though, as liggit points out we need 36209 and 36210.  But it seems like we don't want to do that because somewhere out there somebody might have set a Node Status Condition without intending it for it to do anything.  I maintain that a release note is probably the correct approach there, but it sounds like we're deprecating Node Status Conditions (but keeping them around) and moving to taints instead.

I think there are lessons to be learned from the sequence of events that got us here.  I'm going to write it up.

**Comment by liggitt** [comment]
> somewhere out there somebody might have set a Node Status Condition without intending it for it to do anything

I think it's more likely there are node conditions meant to be seen and acted on by something other than the scheduler

**Comment by davidopp** [comment]
@justinsb I got a bit lost, is there some PR you urgently need for 1.5? I'd rather not do #32609 and #37224 right now but is there something simple you can do to temporarily solve the root problem you are trying to solve?

**Pull request merged**